packages/components/src/internal/components/editable/actions.ts

```diff
@@ -116,75 +96,97 @@ export const loadEditorModelData = async (
+//  same QueryModel to every loader (it uses initEditorModels).
+export const initEditorModel = async (
     queryModel: QueryModel,
-    colFilter?: (col: QueryColumn) => boolean
-): Promise<{ dataModel: QueryModel; editorModel: EditorModel }> => {
-    const response = await loader.fetch(queryModel);
-    const gridData: Partial<QueryModel> = {
-        rows: response.data.toJS(),
-        orderedRows: response.dataIds.toArray(),
-        queryInfo: loader.queryInfo,
-    };
-
-    let columns: QueryColumn[];
+    loader: EditableGridLoader,
+    columnMetadata?: Map<string, EditableColumnMetadata>
+): Promise<EditorModel> => {
+    // TODO: look at EditorModel.getColumns and make sure we're matching the behavior from there as well e.g. file
+    //  columns
+    const { columns: loaderColumns, queryInfo } = loader;
+    // TODO: Most EditableGridLoaders do not actually asynchronously fetch data (see note in EditableGridLoader), we
+    //  can most likely just drop loader as an arg, and have consumers pass in their QueryModel data (if any), and
+    //  skip having QueryModel as an arg.
+    const { data, dataIds } = await loader.fetch(queryModel);
+    // TODO: store rows and orderedRows on EditorModel so we can use it with getUpdatedDataFromEditableGrid
+    const rows = data.toJS();
+    const orderedRows = dataIds.toArray();
     const forUpdate = loader.mode === EditorMode.Update;
-    if (loader.columns) {
-        columns = editorModel.getColumns(
-            gridData.queryInfo,
-            forUpdate,
-            undefined,
-            loader.columns,
-            loader.columns,
-            colFilter
-        );
+    let columns: QueryColumn[];
+
+    if (loaderColumns) {
+        // TODO: investigate how many loaders actually setting columns, there may be a better path forward
+        columns = loader.columns;
+    } else if (forUpdate) {
+        columns = queryInfo.getUpdateColumns();
     } else {
-        columns = editorModel.getColumns(gridData.queryInfo, forUpdate, undefined, undefined, undefined, colFilter);
+        columns = queryInfo.getInsertColumns();
     }
 
-    const editorModelData = await loadEditorModelData(gridData, columns, forUpdate);
+    // file input columns are not supported in the editable grid, so remove them
+    columns = columns.filter(col => !col.isFileInput);
+
+    // Calculate orderedColumns here before we add PK and Container columns to the columns array because they should
+    // be hidden by default.
+    const orderedColumns = columns.map(queryColumn => queryColumn.fieldKey);
+    const columnMap = columns.reduce((result, column) => {
+        result[column.fieldKey] = column;
+        return result;
+    }, {});
+
+    if (forUpdate) {
+        // If we're updating then we need to ensure that the pkCol is in the columnMap so things like readonlyRows
+        // will work.
+        const pkCol = queryInfo.getPkCols()[0];
```

Comment:
QueryInfo isn't guaranteed to have a pkCol. Consider throwing if we do not resolve one in this case.

Reply:
I've since changed this to add all pkCols and not just the first one. That way if we encounter a grid that has multiple it should work (though stuff like readOnlyRows would not be compatible without additional work).